### PR TITLE
feat(eslint-plugin): no input rename rule

### PR DESCRIPTION
**Pull request by sfabriece**

### CODE REVIEW

**Comment by sfabriece** [comment]
I'm struggling with a test case @JamesHenry @mgechev. What is the meaning of `fake renamed`?

```
convertAnnotatedSourceToFailureCase({
      description: 'should fail when an input property is fake renamed',
      annotatedSource: `
      @Component({
        selector: 'foo'
      })
      class TestComponent {
        @Input('foo') label: string;
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      }
          `,
      messageId,
    })
```

**Comment by wKoza** [comment]
You can read these PRs for history : https://github.com/mgechev/codelyzer/pull/561 et https://github.com/mgechev/codelyzer/pull/585 .

`fake renamed` == "renamed unnecessarily"

In your case, the description is wrong. It should be something like "A component shouldn't rename its @input" . It was the rule exposed in the PR [585](https://github.com/mgechev/codelyzer/pull/585) .

**Comment by sfabriece** [comment]
I just copied the description as is from Codelyzer. I'll use yours instead @wKoza .

**Comment by sfabriece** [comment]
rebased master @mgechev

File: packages/eslint-plugin/src/rules/no-input-rename.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+        const propertyAlias = (inputCallExpression
+          .arguments[0] as TSESTree.Literal).value;
+
+        if (
+          propertyAlias &&
+          isIdentifier(classProperty.key) &&
+          canPropertyBeAliased(propertyAlias.toString(), classProperty.key.name)
+        )
+          return;
```
Comment (JamesHenry) [inline packages/eslint-plugin/src/rules/no-input-rename.ts]: Please can you favour using braces for consistency?

Comment (sfabriece) [reply packages/eslint-plugin/src/rules/no-input-rename.ts]: fixed in 4f4c89e

File: packages/eslint-plugin/src/rules/no-input-rename.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        if (inputCallExpression.arguments.length === 0) return;
+
+        let directiveSelectors: ReadonlyArray<any>;
+
+        const canPropertyBeAliased = (
+          propertyAlias: string,
+          propertyName: string,
+        ): boolean => {
+          return !!(
+            (propertyAlias !== propertyName &&
```
[the remark this answers is not in the document]
Comment (sfabriece) [reply packages/eslint-plugin/src/rules/no-input-rename.ts]: I'm not sure of the best way to break this apart, can you offer some guidance @JamesHenry ?

File: packages/eslint-plugin/src/utils/utils.ts [hunk 3]
```diff
@@ -350,3 +350,20 @@ export const getLifecycleInterfaceByMethodName = (
...
...E38\uDE40-\uDE5E\uDED0-\uDEED\uDF00-\uDF2F\uDF40-\uDF43\uDF63-\uDF77\uDF7D-\uDF8F]|\uD81B[\uDF00-\uDF44\uDF50\uDF93-\uDF9F\uDFE0\uDFE1]|\uD821[\uDC00-\uDFEC]|\uD822[\uDC00-\uDEF2]|\uD82C[\uDC00-\uDD1E\uDD70-\uDEFB]|\uD82F[\uDC00-\uDC6A\uDC70-\uDC7C\uDC80-\uDC88\uDC90-\uDC99]|\uD835[\uDC00-\uDC54\uDC56-\uDC9C\uDC9E\uDC9F\uDCA2\uDCA5\uDCA6\uDCA9-\uDCAC\uDCAE-\uDCB9\uDCBB\uDCBD-\uDCC3\uDCC5-\uDD05\uDD07-\uDD0A\uDD0D-\uDD14\uDD16-\uDD1C\uDD1E-\uDD39\uDD3B-\uDD3E\uDD40-\uDD44\uDD46\uDD4A-\uDD50\uDD52-\uDEA5\uDEA8-\uDEC0\uDEC2-\uDEDA\uDEDC-\uDEFA\uDEFC-\uDF14\uDF16-\uDF34\uDF36-\uDF4E\uDF50-\uDF6E\uDF70-\uDF88\uDF8A-\uDFA8\uDFAA-\uDFC2\uDFC4-\uDFCB]|\uD83A[\uDC00-\uDCC4\uDD00-\uDD43]|\uD83B[\uDE00-\uDE03\uDE05-\uDE1F\uDE21\uDE22\uDE24\uDE27\uDE29-\uDE32\uDE34-\uDE37\uDE39\uDE3B\uDE42\uDE47\uDE49\uDE4B\uDE4D-\uDE4F\uDE51\uDE52\uDE54\uDE57\uDE59\uDE5B\uDE5D\uDE5F\uDE61\uDE62\uDE64\uDE67-\uDE6A\uDE6C-\uDE72\uDE74-\uDE77\uDE79-\uDE7C\uDE7E\uDE80-\uDE89\uDE8B-\uDE9B\uDEA1-\uDEA3\uDEA5-\uDEA9\uDEAB-\uDEBB]|\uD869[\uDC00-\uDED6\uDF00-\uDFFF]|\uD86D[\uDC00-\uDF34\uDF40-\uDFFF]|\uD86E[\uDC00-\uDC1D\uDC20-\uDFFF]|\uD873[\uDC00-\uDEA1\uDEB0-\uDFFF]|\uD87A[\uDC00-\uDFE0]|\uD87E[\uDC00-\uDE1D])\S*/g;
```
[the remark this answers is not in the document]
Comment (sfabriece) [reply packages/eslint-plugin/src/utils/utils.ts]: I'm not very familiar with feature detection @JamesHenry. Can you have a look at this?

**Comment by JamesHenry** [comment]
@sfabriece I don't think you gave collaborators the right to update your PR when you opened, so I have had to push the changes to a different branch. Here is the commit: https://github.com/angular-eslint/angular-eslint/commit/87ae9cf9d0a996191e9915c08a4738d09f0ac44d

**Comment by sfabriece** [comment]
@JamesHenry I'm getting an error when I try running `yarn` or `lerna bootstrap` with your changes:

It seems [this line](https://github.com/angular-eslint/angular-eslint/blob/87ae9cf9d0a996191e9915c08a4738d09f0ac44d/packages/eslint-plugin/src/utils/utils.ts#L367) is the problem.

I'm running windows 10 if that is relevant.
```
$ yarn
yarn install v1.17.3
[1/4] Resolving packages...
success Already up-to-date.
$ node ./tools/scripts/apply-patched-microbundle.js && lerna bootstrap && yarn build && lerna link && npm run check-clean-workspace-after-install
lerna notice cli v3.14.1
lerna info bootstrap root only
yarn install v1.17.3
[1/4] Resolving packages...
success Already up-to-date.
$ node ./tools/scripts/apply-patched-microbundle.js && lerna bootstrap && yarn build && lerna link && npm run check-clean-workspace-after-install
lerna notice cli v3.14.1
lerna WARN bootstrap Skipping recursive execution
yarn run v1.17.3
$ lerna run build
lerna notice cli v3.14.1
lerna info Executing command in 4 packages: "yarn run build"
@angular-eslint/eslint-plugin-template: $ rm -rf ./dist && microbundle --no-sourcemap --target=node --compress --format=cjs
@angular-eslint/builder: $ rm -rf ./dist && tsc --project tsconfig.build.json && terser ./dist/index.js --output=./dist/index.js
@angular-eslint/eslint-plugin: $ rm -rf ./dist && microbundle --no-sourcemap --target=node --compress --format=cjs
@angular-eslint/template-parser: $ rm -rf ./dist && microbundle --no-sourcemap --target=node --compress --format=cjs
@angular-eslint/eslint-plugin: node-resolve: setting options.module is deprecated, please override options.mainFields instead
@angular-eslint/eslint-plugin: node-resolve: setting options.jsnext is deprecated, please override options.mainFields instead
@angular-eslint/eslint-plugin-template: node-resolve: setting options.module is deprecated, please override options.mainFields instead
@angular-eslint/eslint-plugin-template: node-resolve: setting options.jsnext is deprecated, please override options.mainFields instead
@angular-eslint/template-parser: node-resolve: setting options.module is deprecated, please override options.mainFields instead
@angular-eslint/template-parser: node-resolve: setting options.jsnext is deprecated, please override options.mainFields instead
@angular-eslint/template-parser: preferring built-in module 'events' over local alternative at 'C:\Users\Fabriece Sumuni\documents\Github\oss\angular-eslint\node_modules\events\events.js', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning
@angular-eslint/template-parser: preferring built-in module 'events' over local alternative at 'C:\Users\Fabriece Sumuni\documents\Github\oss\angular-eslint\node_modules\events\events.js', pass 'preferBuiltins: false' to disable this behavior or 'preferBuiltins: true' to disable this warning
@angular-eslint/template-parser: 'events' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
@angular-eslint/builder: $ cp src/schema.json dist/
@angular-eslint/eslint-plugin-template: Build "eslintPluginTemplate" to dist:
@angular-eslint/eslint-plugin-template:        1200 B: index.js.gz
@angular-eslint/eslint-plugin-template:        1022 B: index.js.br
@angular-eslint/template-parser: Build "templateParser" to dist:
@angular-eslint/template-parser:        1026 B: index.js.gz
@angular-eslint/template-parser:         896 B: index.js.br
@angular-eslint/eslint-plugin: (buble plugin) SyntaxError: Expected atom at position 2
@angular-eslint/eslint-plugin:     \p{L}\S*
@angular-eslint/eslint-plugin:       ^
@angular-eslint/eslint-plugin: at C:/Users/Fabriece Sumuni/documents/Github/oss/angular-eslint/packages/eslint-plugin/src/utils/utils.ts:undefined:undefined
@angular-eslint/eslint-plugin: SyntaxError: Expected atom at position 2
@angular-eslint/eslint-plugin:     \p{L}\S*
@angular-eslint/eslint-plugin:       ^
@angular-eslint/eslint-plugin:     at bail (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regjsparser\parser.js:1106:13)
@angular-eslint/eslint-plugin:     at parseTerm (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regjsparser\parser.js:486:9)
@angular-eslint/eslint-plugin:     at parseAlternative (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regjsparser\parser.js:457:21)
@angular-eslint/eslint-plugin:     at parseDisjunction (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regjsparser\parser.js:437:16)
@angular-eslint/eslint-plugin:     at parse (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regjsparser\parser.js:1121:18)
@angular-eslint/eslint-plugin:     at rewritePattern (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\regexpu-core\rewrite-pattern.js:325:15)
@angular-eslint/eslint-plugin:     at Node.transpile (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\buble\dist\buble.cjs.js:3088:10)
@angular-eslint/eslint-plugin:     at Node.transpile (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\buble\dist\buble.cjs.js:124:11)
@angular-eslint/eslint-plugin:     at Node.transpile (C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\buble\dist\buble.cjs.js:3951:30)
@angular-eslint/eslint-plugin:     at C:\Users\Fabriece Sumuni\Documents\github\oss\angular-eslint\node_modules\buble\dist\buble.cjs.js:3845:16
@angular-eslint/eslint-plugin: error Command failed with exit code 1.
@angular-eslint/eslint-plugin: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run build exited 1 in '@angular-eslint/eslint-plugin'
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
lerna ERR! yarn install --mutex network:42424 --non-interactive exited 1 in '@angular-eslint/angular-eslint'
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

Fabriece Sumuni@Supergirl MINGW64 ~/documents/Github/oss/angular-eslint (no-input-rename)

```

**Comment by JamesHenry** [comment]
Ooops of course we need a try catch, it’s not enough to just check the node version because the code is executed either way. Face palm.

We can either go back to what you had before and just leave an additional note that when we drop Node pre 10 support we can replace the implementation, or can switch to try catch wrapped version.

I cannot say which will be more performant overall for modern node - I would guess the try catch one but need to measure to be sure, I don’t think it’s a big deal either way

File: packages/eslint-plugin/src/utils/utils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgechev) [reply packages/eslint-plugin/src/utils/utils.ts]: @rafaelss95 should have better context here.

[the remark this answers is not in the document]
Comment (rafaelss95) [reply packages/eslint-plugin/src/utils/utils.ts]: Hey @mgechev sry for delay. I've copied this code from @angular's source code (https://github.com/angular/angular/blob/91c7b451d5323f8755b539dc8293deffa9d71cc7/packages/common/src/pipes/case_conversion_pipes.ts#L42-L52).

[the remark this answers is not in the document]
Comment (mgechev) [reply packages/eslint-plugin/src/utils/utils.ts]: The source seems reliable enough.

**Comment by JamesHenry** [comment]
Now that Node 8 is EOL, we could remove the additional checks altogether. WDYT?

**Comment by kimamula** [comment]
Well, can I make a PR for this rule as this PR has been inactive for a while?

https://github.com/angular-eslint/angular-eslint/compare/master...kimamula:no-input-rename

For the `unicodeWordMatch` issue, I cannot understand why that is necessary.
Simply capitalizing the first letter without using RegExp is not enough?
For example, original `no-input-rename` rule in Codelyzer raises an error when the following code is given.
Is this an intended behavior?

```ts
@Directive({
  selector: 'foo'
})
class TestDirective {
  @Input('fooColor') color: string;  // OK
  @Input('fooMycolor') mycolor: string;  // OK
  @Input('fooMyColor') myColor: string;  // NG
}
```

I also applied the rule to setters with `@Input()`.

**Comment by kimamula** [comment]
By the way, is it OK to implement this rule without type checking?

Though the selector `Decorator[expression.callee.name="Input"]` detects decorators whose name is `Input`, it does not guarantee that they are imported from `@angular/core`.

When `parserOptions.project` is set, the type checker and the corresponding node in TypeScript AST can be obtained.

```ts
const typeChecker = context.parserServices.program.getTypeChecker();
const tsNode = context.parserServices.esTreeNodeToTSNodeMap.get(node);
```

What about passing `tsConfig` which is required by `@angular-eslint/builder` to [`eslint.CLIEngine` constructor](https://github.com/angular-eslint/angular-eslint/blob/master/packages/builder/src/index.ts#L199-L203) and use the type checker in the rules which need it?
It also helps to avoid specifying tsconfig.json both in `.eslintrc.js` and `angular.json` (maybe setting it in `.eslintrc.js` and consuming it in `@angular-eslint/builder` is better).

**Comment by JamesHenry** [comment]
@kimamula if you want to open it as a PR we will get it merged. Thanks again @sfabriece for getting things most of the way there!

**Comment by JamesHenry** [comment]
PS I already removed Node 8 support as well

**Pull request closed without merging**